### PR TITLE
Tensor Iterator loop unrolling

### PR DESCRIPTION
**Pull request by jjsjann123**
Summary:

Modified Tensor Iterator gpu reduction kernel.
Creating multiple accumulator during thread reduce, this removes data dependency
between unrolled loops, expose instruction level parallelism that benefits
latency bounded kernels (e.g. welford used by `torch.std`)

This approach increases register usage, such that we need to tune unrolling
factors to prevent register spilling.
Current implementation tune down the unrolling factor to 2 for welford (register
heavy kernel), while keeping it unchanged (4) for the rest of reduction kernels.

### CODE REVIEW

**Comment by jjsjann123** [comment]
This is the loop unrolling PR that was dependent on #17428 

Perf number is listed in https://github.com/pytorch/pytorch/pull/17428#issuecomment-468033996

**Comment by jjsjann123** [comment]
Pinging @umanwizard @colesbury @ngimel for visibility.

**Review by mrshenli — APPROVED** [review]
@pytorchbot retest this please

File: aten/src/ATen/native/cuda/ReduceOpsKernel.cu [hunk 1]
```diff
@@ -25,12 +25,16 @@ void sum_kernel_impl(TensorIterator& iter) {
 
 template <typename scalar_t>
 void std_var_kernel_impl(TensorIterator& iter, bool unbiased, bool take_sqrt) {
-  gpu_reduce_kernel<scalar_t, scalar_t>(iter, WelfordOps<scalar_t, scalar_t, int32_t, float> { unbiased, take_sqrt }, WelfordData<scalar_t, int32_t, float> {});
+  // reducing unrolling factor to 2 for welford kernel
+  // This is necessary to lower register usage that leads to register spills.
+  gpu_reduce_kernel<scalar_t, scalar_t, 2>(iter, WelfordOps<scalar_t, scalar_t, int32_t, float> { unbiased, take_sqrt }, WelfordData<scalar_t, int32_t, float> {});
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply aten/src/ATen/native/cuda/ReduceOpsKernel.cu]: I'll drop some perf number tomorrow.

A calculation to predict register usage by compiler would be very difficult. Given the launch bound we have, each thread can't exceed 32 register usage. Roughly speaking, welfordOps.reduce along would use (4 + 4) = 8 register (https://github.com/pytorch/pytorch/blob/c13ac9b124b240ff20991ab04af27835d14de580/aten/src/ATen/native/SharedReduceOps.h#L51-L63)
An unrolling factor of 4 would take all 32 registers just to hold the value.

**Comment by jjsjann123** [comment]
image
@mrshenli Here's the number with achieved bandwidth

also looks like all tests are passing. I'll push another commit to update the typo in comments (that shouldn't break the test). It should be good to merge then and we don't have to wait for another CI run.

**Comment by mrshenli** [comment]
@jjsjann123 thanks for the numbers, they look great. I noticed that the bandwidth has a big jump between 3200X512 to 6400X512 for the factor=2 case. Do you happen to know the reason behind it?

**Comment by jjsjann123** [comment]
I don't know the precise reason behind the jump. Most likely this means we need to check our launch configs.

But we need to be careful with that, as the same launch configs are also shared with simple reduction kernels. The memory-instruction latency that's hurting welford perf is not relevant for those kernels. Adjustment on the launch config might be tricky to get the best out of both kernels.

**Pull request closed without merging**